### ISSUE TITLE
Standalone (non-MSYS) MinGW-w64 4 and lower do not include TaskDialog() (was: Go get failing)

### ISSUE THREAD

**Issue by vadimyer** [andlabs/ui#147]
```
# github.com/andlabs/ui
C:/GoSpace/src/github.com/andlabs/ui/libui_windows_amd64.a(stddialogs.cpp.obj): In function `msgbox':
E:/github.com/andlabs/libui/windows/stddialogs.cpp:113: undefined reference to `__imp_TaskDialog'
collect2.exe: error: ld returned 1 exit status
```

Windows 10.

**Comment by andlabs** [comment]
Update your MinGW-w64. TaskDialog was only added to comctl32.lib recently.

**Comment by emersion** [comment]
I've the same error with GCC 6.1.1 and MinGW-w64 4.0.6.
```
GOOS=windows GOARCH=amd64 CGO_ENABLED=1 CC=x86_64-w64-mingw32-gcc CXX=x86_64-w64-mingw32-g++ go build
# github.com/andlabs/ui
../../andlabs/ui/libui_windows_amd64.a(stddialogs.cpp.obj): In function `msgbox':
C:\Users\tablet-143\libui\build-windows-amd64/C:/Users/tablet-143/libui/windows/stddialogs.cpp:113: undefined reference to `__imp_TaskDialog'
collect2: error: ld returned 1 exit status
```

**Comment by andlabs** [comment]
Is that the latest standalone version available? o_O It works for me with the MinGW-w64 that is available with MSYS2...

**Comment by emersion** [comment]
Yes, 4.0.6 is the latest. I can see your commit on msys2 (https://github.com/msys2/mingw-w64/commit/5495a11aab96d35f1e9db27f10051412b4190163) but not on the official Git (https://sourceforge.net/p/mingw-w64/mingw-w64/commit_browser). Maybe ask them to apply it?

When I remove the TaskDialog reference, I get the #142 error.

**Comment by omac777** [comment]
On Debian's Sid version updated this morning:
export CC=/usr/bin/x86_64-w64-mingw32-gcc-5.4-win32
export CXX=/usr/bin/x86_64-w64-mingw32-g++-win32
export CGO_ENABLED=1
env GOOS=windows GOARCH=amd64 go build uidemo.go

The compile error I get is:
# github.com/andlabs/ui
../andlabs/ui/libui_windows_amd64.a(stddialogs.cpp.obj): In function `msgbox':
E:\github.com\andlabs\libui\build/E:/github.com/andlabs/libui/windows/stddialogs.cpp:113: undefined reference to `__imp_TaskDialog'
collect2: error: ld returned 1 exit status

I don't recall ever specifying an E: drive on this machine.  Is this coming from the mingw or golang cross-compilers? Or is the "E:" drive coming from andlabs/ui?

I'll try again with msys2 tomorrow morning.

**Comment by andlabs** [comment]
The `E:` is probably coming from MinGW-w64.

**Comment by F0rth** [comment]
here they are https://sourceforge.net/p/mingw-w64/mingw-w64/ci/5495a11aab96d35f1e9db27f10051412b4190163/

**Comment by F0rth** [comment]
So I built a custom mingw-w64 4.0.6 with both 32 and 64 comctl32.def patched, cross compil a basic ui demo for windows/amd64 from a linux/amd64 host and it worked (both on a real windows 8.1/amd64 and with wine).

**Comment by emersion** [comment]
On Arch you can install `mingw-w64-crt-git` from the AUR, it just works.

**Comment by 7sDream** [comment]
@omac777 , Same error here. But I resolve it finally.

---

As @F0rth said, MnGW-w64 version 5.0rc2 added the `TaskDialog` function. 

But because it's a RC release, most Linux haven't upgrade it to 5.0 now.

So we must build it by ourselves for now.

## Download Source Code

The source code of MinGW-w64 5.0rc2 can be download from [here](https://sourceforge.net/projects/mingw-w64/files/mingw-w64/mingw-w64-release/mingw-w64-v5.0-rc2.tar.bz2/download).

## Build and install 5.0rc2 version MinGW-w64 libs

```bash
./configure --prefix=$HOME/some/path --host=x86_64-w64-mingw32
make -j4
make install
```

Replace the `some/path` to a real path where you want put files.

## Let CGO Use the LIBs We Just Build

Add some env var to set the target os, arch, compliers and library path when use go build:

```bash
env GOOS=windows GOARCH=amd64 CGO_ENABLED=1 CXX=x86_64-w64-mingw32-g++ CC=x86_64-w64-mingw32-gcc CGO_LDFLAGS="-L$HOME/some/path/lib" go build -x -o ui.exe ui.go
```

(Notice the `CGO_LDFLAGS` var, the `some/path` is your MinGW-w64 install path)

## Result

Test it with wine:

```bash
wine ui.exe
```

Screenshot:

![](http://rikka-10066868.image.myqcloud.com/713c5b64-dc2b-4c1e-83e1-5a3a8eebc5c4.png)

---

Hope this helps you.

**Comment by omac777** [comment]
Thank you.  It's very helpful.  I'll try it out in the next few days and
let you know.

Cheers.

**Comment by andlabs** [comment]
Renamed the bug report to make it easier to find.

**Comment by lawtonykaps** [comment]
Hi all, been going around in circles because of the error which I get when I try to compile the andlabs ui example. The error is as above. 

C:/Go/src/github.com/andlabs/ui/libui_windows_amd64.a(stddialogs.cpp.obj): In function `msgbox':
E:/github.com/andlabs/libui/windows/stddialogs.cpp:113: undefined reference to `__imp_TaskDialog'
collect2.exe: error: ld returned 1 exit status

Using Windows 10 and Eclipse for the IDE. Sorry but I don't understand what MinGW is or how to install/configure it. Can anyone help me get rid of this problem or does OPEN mean there is not yet any solution.

Many thanks in advance.

**Comment by emersion** [comment]
mingw-w64 5 is needed.

**Comment by lawtonykaps** [comment]
I have searched for mingw-w64 5, downloaded and installed mingw-w64-install.exe.  I now have two new apps. Modify or remove MinGW and a MinGW Command Prompt. There seems to be no way of finding out the installed version.

It's the part I do next that is the problem. Doing the download and install hasn't solved the problem. I have read the forum and some parallel forums but are non the wiser.  Thanks in advance.

**Comment by lawtonykaps** [comment]
...an update. After more reading, I followed the instructions on the site https://github.com/golang/go/wiki/WindowsBuild

Installed the suggested packages and tried to re compile my code. Now I get the following error:
cc1.exe: sorry, unimplemented: 64-bit mode not compiled in

**Comment by emersion** [comment]
This is because you're still using gcc from your old mingw. You can try setting the `CC` environment variable to a 64-bit gcc. It should be prefixed with `mingw-w64-`.

**Comment by lawtonykaps** [comment]
Brilliant. Many thanks. The 64 bit version had already been installed but the Windows 'path' environment variable still contained the 32 bit gcc.exe. Hope these posts helps someone else.

**Comment by andlabs** [comment]
Merged with #279.
